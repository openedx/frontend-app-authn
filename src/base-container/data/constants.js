const IMAGE_LAYOUT = 'image-layout';
const DEFAULT_LAYOUT = 'default-layout';

export { DEFAULT_LAYOUT, IMAGE_LAYOUT };
