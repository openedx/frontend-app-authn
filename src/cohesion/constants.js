export const PAGE_TYPES = {
  ACCOUNT_CREATION: 'account-creation',
  SIGN_IN: 'sign-in',
};

export const ELEMENT_TYPES = {
  BUTTON: 'BUTTON',
};

export const EVENT_TYPES = { ElementClicked: 'redventures.usertracking.v3.ElementClicked' };

export const ELEMENT_TEXT = {
  CREATE_ACCOUNT: 'create-account',
  OPT_IN_TEXT: 'I agree that edx may send me marketing messages',
  SIGN_IN: 'Sign In',
};

export const ELEMENT_NAME = {
  SIGN_IN: PAGE_TYPES.SIGN_IN,
  OPT_OUT: 'opt-out',
  CREATE_ACCOUNT: 'Create an account for free',
};
