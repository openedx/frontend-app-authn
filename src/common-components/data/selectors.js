import { createSelector } from 'reselect';

export const storeName = 'commonComponents';

export const commonComponentsSelector = state => ({ ...state[storeName] });

export const thirdPartyAuthContextSelector = createSelector(
  commonComponentsSelector,
  commonComponents => commonComponents.thirdPartyAuthContext,
);
