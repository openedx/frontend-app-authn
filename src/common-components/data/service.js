import { getConfig } from '@edx/frontend-platform';
import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';

// eslint-disable-next-line import/prefer-default-export
export async function getThirdPartyAuthContext(urlParams) {
  const requestConfig = {
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    params: urlParams,
    isPublic: true,
  };

  const { data } = await getAuthenticatedHttpClient()
    .get(
      `${getConfig().LMS_BASE_URL}/api/mfe_context`,
      requestConfig,
    )
    .catch((e) => {
      throw (e);
    });
  return {
    fieldDescriptions: data.registrationFields || data.registration_fields,
    optionalFields: data.optionalFields || data.optional_fields,
    thirdPartyAuthContext: data.contextData || data.context_data,
  };
}
