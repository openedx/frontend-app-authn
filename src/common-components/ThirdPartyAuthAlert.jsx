import React from 'react';

import { getConfig } from '@edx/frontend-platform';
import { useIntl } from '@edx/frontend-platform/i18n';
import { Alert } from '@edx/paragon';
import PropTypes from 'prop-types';

import messages from './messages';
import { LOGIN_PAGE, REGISTER_PAGE } from '../data/constants';

const ThirdPartyAuthAlert = (props) => {
  const { formatMessage } = useIntl();
  const { currentProvider, referrer } = props;
  const platformName = getConfig().SITE_NAME;
  let message;

  if (referrer === LOGIN_PAGE) {
    message = formatMessage(messages['login.third.party.auth.account.not.linked'], { currentProvider, platformName });
  } else {
    message = formatMessage(messages['register.third.party.auth.account.not.linked'], { currentProvider, platformName });
  }

  if (!currentProvider) {
    return null;
  }

  return (
    <>
      <Alert id="tpa-alert" className={referrer === REGISTER_PAGE ? 'alert-success mt-n2 mb-5' : 'alert-warning mt-n2 mb-5'}>
        {referrer === REGISTER_PAGE ? (
          <Alert.Heading>{formatMessage(messages['tpa.alert.heading'])}</Alert.Heading>
        ) : null}
        <p>{ message }</p>
      </Alert>
      {referrer === REGISTER_PAGE ? (
        <h4 className="mt-4 mb-4">{formatMessage(messages['registration.using.tpa.form.heading'])}</h4>
      ) : null}
    </>
  );
};

ThirdPartyAuthAlert.defaultProps = {
  currentProvider: '',
  referrer: LOGIN_PAGE,
};

ThirdPartyAuthAlert.propTypes = {
  currentProvider: PropTypes.string,
  referrer: PropTypes.string,
};

export default ThirdPartyAuthAlert;
