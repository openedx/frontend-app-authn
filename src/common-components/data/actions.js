import { AsyncActionType } from '../../data/utils';

export const THIRD_PARTY_AUTH_CONTEXT = new AsyncActionType('THIRD_PARTY_AUTH', 'GET_THIRD_PARTY_AUTH_CONTEXT');
export const THIRD_PARTY_AUTH_CONTEXT_CLEAR_ERROR_MSG = 'THIRD_PARTY_AUTH_CONTEXT_CLEAR_ERROR_MSG';

// Third party auth context
export const getThirdPartyAuthContext = (urlParams) => ({
  type: THIRD_PARTY_AUTH_CONTEXT.BASE,
  payload: { urlParams },
});

export const getThirdPartyAuthContextBegin = () => ({
  type: THIRD_PARTY_AUTH_CONTEXT.BEGIN,
});

export const getThirdPartyAuthContextSuccess = (fieldDescriptions, optionalFields, thirdPartyAuthContext) => ({
  type: THIRD_PARTY_AUTH_CONTEXT.SUCCESS,
  payload: { fieldDescriptions, optionalFields, thirdPartyAuthContext },
});

export const getThirdPartyAuthContextFailure = () => ({
  type: THIRD_PARTY_AUTH_CONTEXT.FAILURE,
});

export const clearThirdPartyAuthContextErrorMessage = () => ({
  type: THIRD_PARTY_AUTH_CONTEXT_CLEAR_ERROR_MSG,
});
