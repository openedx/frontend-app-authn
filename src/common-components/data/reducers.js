import { COMPLETE_STATE, PENDING_STATE } from '../../data/constants';
import { THIRD_PARTY_AUTH_CONTEXT, THIRD_PARTY_AUTH_CONTEXT_CLEAR_ERROR_MSG } from './actions';

export const defaultState = {
  fieldDescriptions: {},
  optionalFields: {},
  thirdPartyAuthApiStatus: null,
  thirdPartyAuthContext: {
    currentProvider: null,
    finishAuthUrl: null,
    countryCode: null,
    providers: [],
    secondaryProviders: [],
    pipelineUserDetails: null,
    errorMessage: null,
  },
};

const reducer = (state = defaultState, action = {}) => {
  switch (action.type) {
    case THIRD_PARTY_AUTH_CONTEXT.BEGIN:
      return {
        ...state,
        thirdPartyAuthApiStatus: PENDING_STATE,
      };
    case THIRD_PARTY_AUTH_CONTEXT.SUCCESS: {
      return {
        ...state,
        fieldDescriptions: action.payload.fieldDescriptions.fields,
        optionalFields: action.payload.optionalFields,
        thirdPartyAuthContext: action.payload.thirdPartyAuthContext,
        thirdPartyAuthApiStatus: COMPLETE_STATE,
      };
    }
    case THIRD_PARTY_AUTH_CONTEXT.FAILURE:
      return {
        ...state,
        thirdPartyAuthApiStatus: COMPLETE_STATE,
      };
    case THIRD_PARTY_AUTH_CONTEXT_CLEAR_ERROR_MSG:
      return {
        ...state,
        thirdPartyAuthApiStatus: PENDING_STATE,
        thirdPartyAuthContext: {
          ...state.thirdPartyAuthContext,
          errorMessage: null,
        },
      };
    default:
      return state;
  }
};

export default reducer;
