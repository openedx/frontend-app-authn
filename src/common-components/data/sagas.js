import { logError } from '@edx/frontend-platform/logging';
import { call, put, takeEvery } from 'redux-saga/effects';

import { setCountryFromThirdPartyAuthContext } from '../../register/data/actions';
import {
  getThirdPartyAuthContextBegin,
  getThirdPartyAuthContextFailure,
  getThirdPartyAuthContextSuccess,
  THIRD_PARTY_AUTH_CONTEXT,
} from './actions';
import {
  getThirdPartyAuthContext,
} from './service';

export function* fetchThirdPartyAuthContext(action) {
  try {
    yield put(getThirdPartyAuthContextBegin());
    const {
      fieldDescriptions, optionalFields, thirdPartyAuthContext,
    } = yield call(getThirdPartyAuthContext, action.payload.urlParams);

    yield put(setCountryFromThirdPartyAuthContext(thirdPartyAuthContext.countryCode));
    yield put(getThirdPartyAuthContextSuccess(fieldDescriptions, optionalFields, thirdPartyAuthContext));
  } catch (e) {
    yield put(getThirdPartyAuthContextFailure());
    logError(e);
  }
}

export default function* saga() {
  yield takeEvery(THIRD_PARTY_AUTH_CONTEXT.BASE, fetchThirdPartyAuthContext);
}
