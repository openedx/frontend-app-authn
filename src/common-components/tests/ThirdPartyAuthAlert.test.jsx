import React from 'react';
import renderer from 'react-test-renderer';
import { IntlProvider } from '@edx/frontend-platform/i18n';

import ThirdPartyAuthAlert from '../ThirdPartyAuthAlert';
import { REGISTER_PAGE } from '../../data/constants';

describe('ThirdPartyAuthAlert', () => {
  let props = {};

  beforeEach(() => {
    props = {
      currentProvider: 'Google',
      platformName: 'edX',
    };
  });

  it('should match login page third party auth alert message snapshot', () => {
    const tree = renderer.create(
      <IntlProvider locale="en">
        <ThirdPartyAuthAlert {...props} />
      </IntlProvider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match register page third party auth alert message snapshot', () => {
    props = {
      ...props,
      referrer: REGISTER_PAGE,
    };

    const tree = renderer.create(
      <IntlProvider locale="en">
        <ThirdPartyAuthAlert {...props} />
      </IntlProvider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
