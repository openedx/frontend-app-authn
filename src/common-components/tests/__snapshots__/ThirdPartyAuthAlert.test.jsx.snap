// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThirdPartyAuthAlert should match login page third party auth alert message snapshot 1`] = `
<div
  className="fade alert-content alert-warning mt-n2 alert show"
  id="tpa-alert"
  role="alert"
>
  <div
    className="pgn__alert-message-wrapper"
  >
    <div
      className="alert-message-content"
    >
      <p>
        You have successfully signed into Google, but your Google account does not have a linked Your Platform Name Here account. To link your accounts, sign in now using your Your Platform Name Here password.
      </p>
    </div>
  </div>
</div>
`;

exports[`ThirdPartyAuthAlert should match register page third party auth alert message snapshot 1`] = `
Array [
  <div
    className="fade alert-content alert-success mt-n2 alert show"
    id="tpa-alert"
    role="alert"
  >
    <div
      className="pgn__alert-message-wrapper"
    >
      <div
        className="alert-message-content"
      >
        <div
          className="alert-heading h4"
        >
          Almost done!
        </div>
        <p>
          You've successfully signed into Google! We just need a little more information before you start learning with Your Platform Name Here.
        </p>
      </div>
    </div>
  </div>,
  <h4
    className="mt-4 mb-4"
  >
    Finish creating your account
  </h4>,
]
`;
